File: mirror word.cpp
#include <iostream>
using namespace std;

bool isMirrorPair(char a, char b) {
    if ((a == 'b' && b == 'd') || (a == 'd' && b == 'b')) return true;
    if ((a == 'p' && b == 'q') || (a == 'q' && b == 'p')) return true;
    if (a == b && (a == 's' || a == 'x' || a == 'z' || a == 'o')) return true;
    return false;
}

//bool ismw(string s){
//	int l = 0;
//	int r = s.length()-1;
//	while(l<r){
//		if (!isMirrorPair(s[l], s[r]))
//        	return false;
//        l++;
//        r--;
//	}
//	return true;
//}
bool isMirrorWord(string s, int left, int right) {
    if (left >= right)
        return true;

    if (!isMirrorPair(s[left], s[right]))
        return false;

    return isMirrorWord(s, left + 1, right - 1);
}

int main() {
    string word = "bpoqd";
    cout << isMirrorWord(word, 0, word.length() - 1);
    //cout<< ismw(word);
    return 0;
}

File: ChallengePhase1/MirrorWord.java
package ChallengePhase1;

import java.util.HashMap;
import java.util.Map;

public class MirrorWord {
    public static boolean isMirrorWord(String word , Map<Character,Character> mirrorPairs){
        int length = word.length();
        return checkMirrorWord(word , mirrorPairs , 0 , length-1);
    }
    private static boolean checkMirrorWord(String word , Map<Character,Character> mirrorPairs , int left , int right){
        // Base case
        if(left>right) return true;
        // Check current pointers
        if(!mirrorPairs.containsKey(word.charAt(left)) || mirrorPairs.get(word.charAt(left)) != word.charAt(right)) return false;
        // Recursive call
        return checkMirrorWord(word, mirrorPairs, left+1, right-1);
    }

    public static void main(String[] args) {
        Map<Character,Character> mirrorPairs = new HashMap<>();
        mirrorPairs.put('b', 'd');
        mirrorPairs.put('d', 'b');
        mirrorPairs.put('p', 'q');
        mirrorPairs.put('q', 'p');
        mirrorPairs.put('s', 's');
        mirrorPairs.put('x', 'x');
        mirrorPairs.put('z', 'z');
        mirrorPairs.put('o', 'o');

        System.out.println(isMirrorWord("oofoo" , mirrorPairs));

    }
}
